test/distributed/_composable/fsdp/test_fully_shard_state_dict.py

```diff
@@ -0,0 +1,157 @@
+# Owner(s): ["oncall: distributed"]
+
+import copy
+import functools
+
+from typing import Dict
+
+import torch
+import torch.nn as nn
+from torch.distributed._composable.fsdp import fully_shard
+from torch.distributed._tensor import DTensor
+from torch.distributed.device_mesh import DeviceMesh, init_device_mesh
+from torch.distributed.tensor.parallel import (
+    ColwiseParallel,
+    parallelize_module,
+    RowwiseParallel,
+)
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import FSDPTest, MLP
+from torch.testing._internal.common_utils import run_tests
+
+
+class TestFullyShardStateDict(FSDPTest):
```

Comment:
I only included some basic state dict tests. For example, the 2D state dict tests here do not break even without having strided sharding.

The tests are written mainly to exercise the state dict hooks to force-reshard parameters if saving or loading a state dict.

Reply:
> I only included some basic state dict tests. For example, the 2D state dict tests here do not break even without having strided sharding.
> 
> The tests are written mainly to exercise the state dict hooks to force-reshard parameters if saving or loading a state dict.

2D state_dict should be working right now because you are sharding the tensors from the parent_mesh directly. We can do this because we know that FSDP is the outermost dimension and the state_dict call should come from the outermost parallelism. 

Ideally, the flow would be a given parallelism should not need to inquire the parent_mesh to produce the globally sound DTensor.  This would be more composable friendly, as then in theory the order of parallelism would not matter. To achieve this flexibility, I believe we will need the `distribute_tensor` API to accept DTensor and that may require the strided sharding that @wanchaol mentioned.